File: cprop/prop/original.ml
open GT
open OCanren
open OCanren.Std
open Helper

let topLevel x0 = 
  let rec fail () = fail 
  and ando x y b = (((x === (trueo ())) &&& (y === (trueo ())) &&& (b === (trueo ()))) ||| ((x === (falso ())) &&& (y === (trueo ())) &&& (b === (falso ()))) ||| ((x === (trueo ())) &&& (y === (falso ())) &&& (b === (falso ()))) ||| ((x === (falso ())) &&& (y === (falso ())) &&& (b === (falso ())))) 
  and oro x y b = (((x === (trueo ())) &&& (y === (trueo ())) &&& (b === (trueo ()))) ||| ((x === (falso ())) &&& (y === (trueo ())) &&& (b === (trueo ()))) ||| ((x === (trueo ())) &&& (y === (falso ())) &&& (b === (trueo ()))) ||| ((x === (falso ())) &&& (y === (falso ())) &&& (b === (falso ())))) 
  and noto x b = (((x === (trueo ())) &&& (b === (falso ()))) ||| ((x === (falso ())) &&& (b === (trueo ())))) 
  and implicationo x y b = (((x === (falso ())) &&& (y === (trueo ())) &&& (b === (trueo ()))) ||| ((x === (falso ())) &&& (y === (falso ())) &&& (b === (trueo ()))) ||| ((x === (trueo ())) &&& (y === (trueo ())) &&& (b === (trueo ()))) ||| ((x === (trueo ())) &&& (y === (falso ())) &&& (b === (falso ())))) 
  and evalo st fm u = (fresh (x y v w z) (((fm === (lit u)) ||| ((fm === (var z)) &&& (elemo z st u)) ||| ((noto v u) &&& (evalo st x v) &&& (fm === (neg x))) ||| ((oro v w u) &&& (evalo st x v) &&& (evalo st y w) &&& (fm === (disj x y))) ||| ((ando v w u) &&& (evalo st x v) &&& (evalo st y w) &&& (fm === (conj x y))) ||| ((implicationo v w u) &&& (evalo st x v) &&& (evalo st y w) &&& (fm === (impl x y)))))) 
  and elemo n s v = (fresh (h t n') ((((n === (zero ())) &&& (s === (h % t)) &&& (v === h)) ||| ((s === (h % t)) &&& (elemo n' t v) &&& (n === (succ n')))))) 
  in        (fresh (x) ((evalo ((List.nil ())) ((disj x x)) ((trueo ())))))

File: res/prop/consPD/original.pl
fail() :- fail().
ando(trueo, trueo, trueo).
ando(falso, trueo, falso).
ando(trueo, falso, falso).
ando(falso, falso, falso).
oro(trueo, trueo, trueo).
oro(falso, trueo, trueo).
oro(trueo, falso, trueo).
oro(falso, falso, falso).
noto(trueo, falso).
noto(falso, trueo).
implicationo(falso, trueo, trueo).
implicationo(falso, falso, trueo).
implicationo(trueo, trueo, trueo).
implicationo(trueo, falso, falso).
evalo(St, lit(U), U).
evalo(St, var(Z), U) :- elemo(Z, St, U).
evalo(St, disj(X, Y), U) :- oro(V, W, U), evalo(St, X, V), evalo(St, Y, W).
evalo(St, conj(X, Y), U) :- ando(V, W, U), evalo(St, X, V), evalo(St, Y, W).
elemo(zero, cons(H, T), H).
elemo(succ(N_0), cons(H, T), V) :- elemo(N_0, T, V).